packages/grid/_modules_/grid/hooks/features/useComponents.tsx

```diff
@@ -1,93 +1,66 @@
 import * as React from 'react';
-import { ComponentProps, ApiRef, GridComponentOverridesProp, RootContainerRef } from '../../models';
-import { ErrorMessage } from '../../components/ErrorMessage';
-import { LoadingOverlay } from '../../components/LoadingOverlay';
-import { NoRowMessage } from '../../components/NoRowMessage';
-import { optionsSelector } from '../utils/useOptionsProp';
+import { DefaultFooterProps } from '../../components/DefaultFooter';
+import { GridColumnHeaderMenuItemProps } from '../../components/menu/columnMenu/GridColumnHeaderMenu';
+import {
+  BaseComponentProps,
+  ApiRef,
+  GridSlotsComponent,
+  RootContainerRef,
+  DEFAULT_SLOTS_COMPONENTS,
+} from '../../models';
+import { ErrorMessageProps } from '../../components/ErrorMessage';
+import { optionsSelector } from '../utils/optionsSelector';
 import { visibleColumnsSelector } from './columns/columnsSelector';
 import { useGridSelector } from './core/useGridSelector';
-import { paginationSelector } from './pagination/paginationSelector';
+import { useGridState } from './core/useGridState';
 import { unorderedRowModelsSelector } from './rows/rowsSelector';
 
 export const useComponents = (
-  componentOverrides: GridComponentOverridesProp | undefined,
+  componentsProp: GridSlotsComponent | undefined,
   apiRef: ApiRef,
   gridRootRef: RootContainerRef,
 ) => {
   const options = useGridSelector(apiRef, optionsSelector);
   const rows = useGridSelector(apiRef, unorderedRowModelsSelector);
   const columns = useGridSelector(apiRef, visibleColumnsSelector);
-  const pagination = useGridSelector(apiRef, paginationSelector);
+  const [state] = useGridState(apiRef!);
 
-  const componentProps: ComponentProps = React.useMemo(
+  const baseComponentProps: BaseComponentProps = React.useMemo(
     () => ({
-      pagination,
+      state,
       rows,
       columns,
       options,
       api: apiRef,
       rootElement: gridRootRef,
     }),
-    [pagination, rows, columns, options, apiRef, gridRootRef],
+    [state, rows, columns, options, apiRef, gridRootRef],
   );
 
-  const headerComponent = React.useMemo(
-    () =>
-      componentOverrides?.header
-        ? React.createElement(componentOverrides.header, componentProps)
-        : null,
-    [componentOverrides, componentProps],
-  );
+  const components = React.useMemo(() => {
+    const allComponents = { ...DEFAULT_SLOTS_COMPONENTS, ...componentsProp };
```

Comment:
I'm not sure. This wouldn't work:

```jsx
<DataGrid components={{ Footer: foo ? MyGridFooter : undefined }} />
```

Reply:
using undefined is not ideal as 
```
const componentsA = {}  
const componentsB = {Footer: undefined} // it should be null
componentsA.Footer === componentsB.Footer => true

```

Reply:
It would be same issue with null

<DataGrid components={{ Footer: foo ? MyGridFooter : null }} />

Reply:
Anyway if you define the prop then `undefined` will override the default value